src/input_common/joycon/jc_adapter.cpp

```diff
@@ -0,0 +1,372 @@
+        if (result < 1) {
+            LOG_ERROR(Input, "No response from joystick");
+        }
+        tries++;
+    } while (tries < 10 && buffer[0] != 0x21 && buffer[14] != sc);
+    return buffer;
+}
+
+u8 Joycons::get_counter() {
+    u8 counter = global_counter;
+    if (global_counter == 0x0F) {
+        global_counter = 0;
+    } else {
+        ++global_counter;
+    }
+    return counter;
+}
+
+JCPadStatus Joycons::GetLeftPadButtons(std::size_t port, u8* buffer) {
+    JCPadStatus pad = {};
+
+    adapter_controllers_status[port] = JoyControllerTypes::Left;
+
+    static constexpr std::array<PadButton, 8> b1_buttons{
+        PadButton::PAD_BUTTON_A,    PadButton::PAD_BUTTON_B,    PadButton::PAD_BUTTON_X,
+        PadButton::PAD_BUTTON_Y,    PadButton::PAD_BUTTON_LEFT, PadButton::PAD_BUTTON_RIGHT,
+        PadButton::PAD_BUTTON_DOWN, PadButton::PAD_BUTTON_UP,
+    };
+
+    static constexpr std::array<PadButton, 4> b2_buttons{
+        PadButton::PAD_BUTTON_START,
+        PadButton::PAD_TRIGGER_Z,
+        PadButton::PAD_TRIGGER_R,
+        PadButton::PAD_TRIGGER_L,
+    };
+
+    static constexpr std::array<PadAxes, 6> axes{
+        PadAxes::StickX,    PadAxes::StickY,      PadAxes::SubstickX,
+        PadAxes::SubstickY, PadAxes::TriggerLeft, PadAxes::TriggerRight,
+    };
+
+    if (adapter_controllers_status[port] == JoyControllerTypes::None && !get_origin[port]) {
+        // Controller may have been disconnected, recalibrate if reconnected.
+        get_origin[port] = true;
+    }
+
+    if (adapter_controllers_status[port] != JoyControllerTypes::None) {
+        const u8 b1 = buffer[5];
+        const u8 b2 = 0;
+
+        for (std::size_t i = 0; i < b1_buttons.size(); ++i) {
+            if ((b1 & (1U << i)) != 0) {
+                pad.button |= static_cast<u16>(b1_buttons[i]);
+            }
+        }
+
+        for (std::size_t j = 0; j < b2_buttons.size(); ++j) {
+            if ((b2 & (1U << j)) != 0) {
+                pad.button |= static_cast<u16>(b2_buttons[j]);
+            }
+        }
+
+        for (PadAxes axis : axes) {
+            const std::size_t index = static_cast<std::size_t>(axis);
+            pad.axis_values[index] = 0;
+        }
+
+        if (get_origin[port]) {
+            origin_status[port].axis_values = pad.axis_values;
+            get_origin[port] = false;
+        }
+    }
+    return pad;
+}
+
+void Joycons::PadToState(const JCPadStatus& pad, JCState& state) {
+    for (const auto& button : PadButtonArray) {
+        const u16 button_value = static_cast<u16>(button);
+        state.buttons.insert_or_assign(button_value, pad.button & button_value);
+    }
+
+    for (size_t i = 0; i < pad.axis_values.size(); ++i) {
+        state.axes.insert_or_assign(static_cast<u8>(i), pad.axis_values[i]);
+    }
+}
+
+void Joycons::Read() {
+    LOG_DEBUG(Input, "JC Adapter Read() thread started");
+
+    u8* buffer = new u8[report_len];
```

Comment:
```suggestion
    std::vector<u8> buffer(report_len);
```

The for loop below this can then be removed.